cpp/src/bitmask/bit_mask.cuh

```diff
@@ -14,143 +14,224 @@
  * limitations under the License.
  */
 
-#ifndef _BIT_MASK_H_
-#define _BIT_MASK_H_
+#ifndef CUDF_BIT_MASK_CUH
+#define CUDF_BIT_MASK_CUH
 
-#include "cudf.h"
-#include "bit_mask.h"
-#include "rmm/rmm.h"
-#include "utilities/error_utils.hpp"
-#include <cuda_runtime_api.h>
+#include <utilities/cudf_utils.h>
+#include <utilities/bit_util.cuh>
+#include <utilities/integer_utils.hpp>
+#include <utilities/error_utils.hpp>
 
-using bit_mask_t = bit_mask::bit_mask_t;
-
-
-/* ---------------------------------------------------------------------------- *
- * @brief  Class for managing bit containers on the device
- * ---------------------------------------------------------------------------- */
-class BitMask {
-public:
-   __host__ __device__ BitMask(bit_mask_t *valid, int bitlength): valid_(valid), bitlength_(bitlength) {}
+namespace bit_mask {
+  typedef uint32_t     bit_mask_t;
+  enum { bits_per_element = gdf::util::size_in_bits<bit_mask_t>() };
 
   /**
-   * @brief Check to see if a record is Valid (aka not null)
+   * @brief determine the number of bit_mask_t elements are used
    *
-   * @param[in] record_idx   the record index to check
+   * @param[in]  size    Number of bits in the bitmask
    *
-   * @return  true if record is valid, false if record is null
+   * @return the number of elements
    */
-  template <typename T>
-  __device__ bool is_valid(T record_idx) const {
-    return bit_mask::is_valid(valid_, record_idx);
+  CUDA_HOST_DEVICE_CALLABLE
+  constexpr gdf_size_type num_elements(gdf_size_type size) {
+//    return (( size + ( bits_per_element - 1)) / bits_per_element );
```

Comment:
I only changed the functions in this file to use code from bit_util.cuh - I didn't change what they actually do. My reasoning was to limit the scope of my changes; but then - I was making those changes as a detour from other work.

To be honest, I'm not sure it even makes sense to have this file and `bit_util.cuh`. This also relates to issue #961 .

Now, I can do one of two things: 

1. Address your specific comments/suggestions only
2. Go deeper by:
    2.1 removing any redundancies between the code here and `bit_util.cuh`
    2.2 removing code here that's not used anywhere else
    2.3 moving code here that's only used in the BitMask class into the `protected` section of that class

What would you prefer?

Reply:
@harrism : Ping. I'll need feedback from you here to move this PR forward. Sorry for force-push last week (it was before you asked me not to do those).

Reply:
I don't know. Sorry, this whole PR has me confused, and it's been a while since I looked at it.

Someone did a PR for us in the last release to create a new Bitmask class that is sitting unused. I'd rather see use of that class be put into place.

Reply:
> I'd rather see use of that class be put into place.

You probably wouldn't. If we're talking about the `BitMask` class - it's quite half-baked, plus it makes assumptions it shouldn't be (mostly about the basic bit-container unit size, which should be a template parameter).

This PR is not the final form of how things could and should look, but it _is_ a significant improvement. In the (hopefully near) future I intend to write a bit_span class which would be like a span for bits, with nice proxies like in vector<bool>, iterators, the works. But for now it's better for this to go in.

Reply:
Please file issues before filing PRs so we can discuss whether we want the changes.

Reply:
@harrism : 

1. Ok, will do so in the future in general; do you also mean issues relating to _this_ PR?
2. I'm also willing to invest time and effort in reviewing PRs  
3. Remember you asked me to try to limit the amount of auxiliary code in my apply_boolean_mask PR; this motivated me to propose several PRs including this one.
4. At least two other actual-feature PRs of mine depend on this one. Specifically, the PR for a multisearch API function.